smart/IGM.py

```diff
@@ -0,0 +1,116 @@
+import matplotlib.pyplot as plt
+import numpy as np
+import skimage as ski
+from skimage.morphology import disk
+
+import astropy.units as u
+
+from sunpy.map import Map
+
+from smart.map_processing import STL
+
+
+def IGM(im_map: Map, diff_map: Map, dilation_radius: u.Quantity[u.arcsec] = 2.5 * u.arcsec):
+    """
+    Performing Indexing and Growing of the Mask (hence the name IGM).
+
+    Transient features are removed by comparing the mask at time 't' and the mask differentially
+    rotated to time 't'. ARs are then assigned ascending integer values (starting from one) in
+    order of decreasing size.
+
+    Parameters
+    ----------
+    im_map : Map
+        Processed magnetogram map from time 't'.
+    diff_map : Map
+        Processed magnetogtam map from time 't - delta_t' differentially rotated to time t.
+    dilation_radius : int, optional
+        Radius of the disk for binary dilation (default is 2.5 arcsecs).
+
+    Returns
+    -------
+    sorted_labels : numpy.ndarray
+        Individual contiguous features are indexed by assigning ascending integer
+        values (beginning with one) in order of decreasing feature size.
+
+    """
+    arcsec_to_pixel = ((im_map.scale[0] + im_map.scale[1]) / 2) ** (-1)
+    dilation_radius = (np.round(dilation_radius * arcsec_to_pixel)).to_value(u.pix)
+
+    filtered_labels = STL(im_map)[1]
+    filtered_labels_dt = STL(diff_map)[1]
+
+    dilated_mask = ski.morphology.binary_dilation(filtered_labels, disk(dilation_radius))
+    dilated_mask_dt = ski.morphology.binary_dilation(filtered_labels_dt, disk(dilation_radius))
+
+    transient_features = dilated_mask_dt & ~dilated_mask
+    final_mask = dilated_mask & ~transient_features
+    final_labels = ski.measure.label(final_mask)
+
+    regions = ski.measure.regionprops(final_labels)
+    region_sizes = [(region.label, region.area) for region in regions]
+
+    sorted_region_sizes = sorted(region_sizes, key=lambda x: x[1], reverse=True)
+    sorted_labels = np.zeros_like(final_labels)
+    for new_label, (old_label, _) in enumerate(sorted_region_sizes, start=1):
+        sorted_labels[final_labels == old_label] = new_label
+
+    return sorted_labels
+
+
+def plot_IGM(im_map: Map, diff_map: Map, contours=True, labels=True, figtext=True):
```

Comment:
Would be better to to not run the IGM in here but pass in the results e.g.

```
regions = IGM(...)
plot_regions(region)
```
I know this might mean the function has to return more variables but that's ok.